src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,412 @@
+{
+    internal class PreloginPacketHandler : IHandler<PreLoginHandlerContext>
+    {
+        const int GUID_SIZE = 16;
+
+        const int PAYLOAD_OFFSET_AND_LENGTH_SIZE_IN_BYTES = 4;
+
+        const int PAYLOAD_LENGTH_SIZE_IN_BYTES = 2;
+
+        // EventSource counter
+        private static int s_objectTypeCount;
+
+        internal readonly int _objectID = Interlocked.Increment(ref s_objectTypeCount);
+
+        internal int ObjectID => _objectID;
+
+        public IHandler<PreLoginHandlerContext> NextHandler { get; set; }
+
+        public async ValueTask Handle(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            await PreloginPacketHandler.CreatePreLoginAndSend(context, isAsync, ct).ConfigureAwait(false);
+            if (context.HasError)
+            {
+                return;
+            }
+            await ReadPreLoginresponse(context, isAsync, ct).ConfigureAwait(false);
+
+            if (context.HasError)
+            {
+                return;
+            }
+
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Consume the Prelogin response from the server.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async Task ReadPreLoginresponse(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            context.ConnectionContext.MarsCapable = context.ConnectionContext.ConnectionString.MARS; // Assign default value
+            context.ConnectionContext.FedAuthRequired = false;
+
+            TdsStream tdsStream = context.ConnectionContext.TdsStream;
+
+            // Look into the first byte to see if we are connecting to a 6.5 or earlier server.
+            int option = await tdsStream.ReadByteAsync(isAsync, ct).ConfigureAwait(false);
+
+            if (option == 0xaa)
+            {
+                // If the first byte is 0xAA, we are connecting to a 6.5 or earlier server, which
+                // is not supported.
+                throw SQL.InvalidSQLServerVersionUnknown();
+            }
+
+            byte[] payload = new byte[tdsStream.PacketDataLeft];
+
+            int payloadOffset = 0;
+            int offset = 0;
+
+            while (option != (byte)PreLoginOptions.LASTOPT)
+            {
+                switch (option)
+                {
+                    case (int)PreLoginOptions.VERSION:
+
+                        payloadOffset = payload[offset++] << 8 | payload[offset++];
+                        offset += PAYLOAD_LENGTH_SIZE_IN_BYTES; // Skip the payload length
+
+                        byte majorVersion = payload[payloadOffset];
+
+                        bool is2005OrLater = majorVersion >= 9;
```

Comment:
Create and use constant e.g. MIN_SQLSERVER_MAJOR_VERSION, than literal '9'

Reply:
Removed from code.